File: maki-code-index/src/csharp.rs
use tree_sitter::Node;

use crate::common::{
    ChildKind, FIELD_TRUNCATE_THRESHOLD, LanguageExtractor, Section, SkeletonEntry, compact_ws,
    find_child, line_range, node_text, prefixed,
};

pub(crate) struct CSharpExtractor;

const MODIFIER_KEYWORDS: &[&str] = &[
    "public",
    "private",
    "protected",
    "internal",
    "static",
    "abstract",
    "sealed",
    "override",
    "virtual",
    "async",
    "readonly",
    "extern",
    "partial",
    "new",
    "unsafe",
    "volatile",
];

impl CSharpExtractor {
    fn modifiers_text(&self, node: Node, source: &[u8]) -> String {
        let mut parts = Vec::new();
        let mut cursor = node.walk();
        for child in node.children(&mut cursor) {
            let text = node_text(child, source);
            let include = match child.kind() {
                "modifier" => MODIFIER_KEYWORDS.contains(&text),
                "attribute_list" => true,
                _ => false,
            };
            if include {
                parts.push(text.to_string());
            }
        }
        parts.join(" ")
    }

    fn base_list_text(&self, node: Node, source: &[u8]) -> String {
        let Some(bl) = find_child(node, "base_list") else {
            return String::new();
        };
        let text = node_text(bl, source);
        let trimmed = text.trim_start_matches(':').trim();
        format!(" : {trimmed}")
    }

    fn extract_import(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let text = node_text(node, source);
        let cleaned = text
            .strip_prefix("using ")
            .unwrap_or(text)
            .trim_end_matches(';')
            .trim();
        let paths = vec![cleaned.split('.').map(String::from).collect()];
        Some(SkeletonEntry::new_import(node, paths))
    }

    fn extract_namespace(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        Some(SkeletonEntry::new(Section::Module, node, name.to_string()))
    }

    fn extract_class(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let mods = self.modifiers_text(node, source);
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        let bases = self.base_list_text(node, source);
        let label = prefixed(&mods, format_args!("class {name}{bases}"));
        let children = self.extract_declaration_list(node, source);
        Some(SkeletonEntry::new(Section::Class, node, label).with_children(children))
    }

    fn extract_struct(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let mods = self.modifiers_text(node, source);
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        let bases = self.base_list_text(node, source);
        let label = prefixed(&mods, format_args!("struct {name}{bases}"));
        let children = self.extract_declaration_list(node, source);
        Some(SkeletonEntry::new(Section::Type, node, label).with_children(children))
    }

    fn extract_interface(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let mods = self.modifiers_text(node, source);
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        let bases = self.base_list_text(node, source);
        let label = prefixed(&mods, format_args!("interface {name}{bases}"));
        let children = self.extract_interface_body(node, source);
        Some(SkeletonEntry::new(Section::Trait, node, label).with_children(children))
    }

    fn extract_record(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let mods = self.modifiers_text(node, source);
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        let params = find_child(node, "parameter_list")
            .map(|n| node_text(n, source))
            .unwrap_or("");
        let bases = self.base_list_text(node, source);
        let label = prefixed(&mods, format_args!("record {name}{params}{bases}"));
        Some(SkeletonEntry::new(Section::Type, node, label))
    }

    fn extract_enum(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let mods = self.modifiers_text(node, source);
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        let label = prefixed(&mods, format_args!("enum {name}"));
        let body = node.child_by_field_name("body")?;
        let mut constants = Vec::new();
        let mut cursor = body.walk();
        for child in body.children(&mut cursor) {
            if child.kind() == "enum_member_declaration" {
                let cname = child
                    .child_by_field_name("name")
                    .map(|n| node_text(n, source))
                    .unwrap_or("_");
                constants.push(cname.to_string());
            }
        }
        Some(
            SkeletonEntry::new(Section::Type, node, label)
                .with_children(constants)
                .with_child_kind(ChildKind::Brief),
        )
    }

    fn extract_declaration_list(&self, node: Node, source: &[u8]) -> Vec<String> {
        let Some(body) = node.child_by_field_name("body") else {
            return Vec::new();
        };
        let mut members = Vec::new();
        let mut field_count = 0usize;
        let mut cursor = body.walk();
        for child in body.children(&mut cursor) {
            match child.kind() {
                "method_declaration" | "constructor_declaration" => {
                    let sig = self.method_signature(child, source);
                    let lr =
                        line_range(child.start_position().row + 1, child.end_position().row + 1);
                    members.push(format!("{sig} {lr}"));
                }
                "field_declaration" => {
                    field_count += 1;
                    if field_count <= FIELD_TRUNCATE_THRESHOLD {
                        let text = self.field_text(child, source);
                        let lr = line_range(
                            child.start_position().row + 1,
                            child.end_position().row + 1,
                        );
                        members.push(format!("{text} {lr}"));
                    }
                }
                "property_declaration" => {
                    field_count += 1;
                    if field_count <= FIELD_TRUNCATE_THRESHOLD {
                        let text = self.property_text(child, source);
                        let lr = line_range(
                            child.start_position().row + 1,
                            child.end_position().row + 1,
                        );
                        members.push(format!("{text} {lr}"));
                    }
                }
                _ => {}
            }
        }
        if field_count > FIELD_TRUNCATE_THRESHOLD {
            members.push("...".into());
        }
        members
    }

    fn extract_interface_body(&self, node: Node, source: &[u8]) -> Vec<String> {
        let Some(body) = node.child_by_field_name("body") else {
            return Vec::new();
        };
        let mut members = Vec::new();
        let mut cursor = body.walk();
        for child in body.children(&mut cursor) {
            let text = match child.kind() {
                "method_declaration" => self.method_signature(child, source),
                "property_declaration" => self.property_text(child, source),
                _ => continue,
            };
            let lr = line_range(child.start_position().row + 1, child.end_position().row + 1);
            members.push(format!("{text} {lr}"));
        }
        members
    }

    fn method_signature(&self, node: Node, source: &[u8]) -> String {
        let mods = self.modifiers_text(node, source);
        let ret = node
            .child_by_field_name("returns")
            .or_else(|| node.child_by_field_name("type"))
            .map(|n| node_text(n, source))
            .unwrap_or("");
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))
            .unwrap_or("_");
        let params = node
            .child_by_field_name("parameters")
            .map(|n| node_text(n, source))
            .unwrap_or("()");
        let base = if ret.is_empty() {
            format!("{name}{params}")
        } else {
            format!("{ret} {name}{params}")
        };
        compact_ws(&prefixed(&mods, format_args!("{base}")))
    }

    fn field_text(&self, node: Node, source: &[u8]) -> String {
        let mods = self.modifiers_text(node, source);
        let decl = find_child(node, "variable_declaration");
        let ty = decl
            .and_then(|n| n.child_by_field_name("type"))
            .map(|n| node_text(n, source))
            .unwrap_or("_");
        let name = decl
            .and_then(|d| find_child(d, "variable_declarator"))
            .and_then(|n| n.child_by_field_name("name"))
            .map(|n| node_text(n, source))
            .unwrap_or("_");
        prefixed(&mods, format_args!("{ty} {name}"))
    }

    fn property_text(&self, node: Node, source: &[u8]) -> String {
        let mods = self.modifiers_text(node, source);
        let ty = node
            .child_by_field_name("type")
            .map(|n| node_text(n, source))
            .unwrap_or("_");
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))
            .unwrap_or("_");
        prefixed(&mods, format_args!("{ty} {name}"))
    }
}

impl LanguageExtractor for CSharpExtractor {
    fn extract_nodes(&self, node: Node, source: &[u8], _attrs: &[Node]) -> Vec<SkeletonEntry> {
        match node.kind() {
            "using_directive" => self.extract_import(node, source).into_iter().collect(),
            "namespace_declaration" | "file_scoped_namespace_declaration" => {
                self.extract_namespace(node, source).into_iter().collect()
            }
            "class_declaration" => self.extract_class(node, source).into_iter().collect(),
            "struct_declaration" => self.extract_struct(node, source).into_iter().collect(),
            "interface_declaration" => self.extract_interface(node, source).into_iter().collect(),
            "enum_declaration" => self.extract_enum(node, source).into_iter().collect(),
            "record_declaration" => self.extract_record(node, source).into_iter().collect(),
            _ => Vec::new(),
        }
    }

    fn is_doc_comment(&self, node: Node, source: &[u8]) -> bool {
        node.kind() == "single_line_doc_comment"
            || (node.kind() == "comment" && node_text(node, source).starts_with("///"))
    }

    fn import_separator(&self) -> &'static str {
        "."
    }
}

File: plugins/index/lang/c_sharp.lua
return function(U)
  local get_text = U.get_text
  local find_child = U.find_child
  local compact_ws = U.compact_ws
  local new_entry = U.new_entry
  local new_import_entry = U.new_import_entry
  local simple_import = U.simple_import
  local SECTION = U.SECTION
  local CHILD_BRIEF = U.CHILD_BRIEF
  local prefixed = U.prefixed
  local extract_enum_variants = U.extract_enum_variants
  local extract_body_members = U.extract_body_members
  local format_range = U.format_range
  local line_start = U.line_start
  local line_end = U.line_end

  local MODIFIER_KEYWORDS = {
    public = true, private = true, protected = true, internal = true,
    static = true, abstract = true, sealed = true, override = true,
    virtual = true, async = true, readonly = true, extern = true,
    partial = true, new = true, unsafe = true, volatile = true,
  }

  local function modifiers_text_free(node, source)
    local parts = {}
    for _, child in ipairs(node:children()) do
      local ckind = child:type()
      if ckind == "modifier" then
        local t = get_text(child, source)
        if MODIFIER_KEYWORDS[t] then parts[#parts + 1] = t end
      elseif ckind == "attribute_list" then
        parts[#parts + 1] = get_text(child, source)
      end
    end
    return table.concat(parts, " ")
  end

  local function base_list_text(node, source)
    local bl = find_child(node, "base_list")
    if not bl then return "" end
    local t = get_text(bl, source):match("^%s*(.-)%s*$")
    t = t:gsub("^:", ""):match("^%s*(.-)%s*$")
    return " : " .. t
  end

  local function method_signature_free(node, source)
    local mods = modifiers_text_free(node, source)
    local ret = node:field("returns")[1] or node:field("type")[1]
    local ret_str = ret and (get_text(ret, source) .. " ") or ""
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local params_node = find_child(node, "parameter_list")
    local params = params_node and get_text(params_node, source) or "()"
    return compact_ws(prefixed(mods, ret_str .. name .. params))
  end

  local function field_text_free(node, source)
    local mods = modifiers_text_free(node, source)
    local decl = find_child(node, "variable_declaration")
    if not decl then return compact_ws(prefixed(mods, get_text(node, source))) end
    local type_node = decl:field("type")[1]
    local type_str = type_node and get_text(type_node, source) or "_"
    local names = {}
    for _, child in ipairs(decl:children()) do
      if child:type() == "variable_declarator" then
        local nn = child:field("name")[1]
        if nn then names[#names + 1] = get_text(nn, source) end
      end
    end
    local name_str = #names > 0 and table.concat(names, ", ") or "_"
    return compact_ws(prefixed(mods, type_str .. " " .. name_str))
  end

  local function property_text_free(node, source)
    local mods = modifiers_text_free(node, source)
    local type_node = node:field("type")[1]
    local type_str = type_node and (get_text(type_node, source) .. " ") or ""
    local name_node = node:field("name")[1]
    local name = name_node and get_text(name_node, source) or "_"
    return compact_ws(prefixed(mods, type_str .. name))
  end

  local CLASS_RULES = {
    { kind = "method_declaration",      handler = "method", fn = method_signature_free },
    { kind = "constructor_declaration", handler = "method", fn = method_signature_free },
    { kind = "field_declaration",       handler = "field",  fn = field_text_free,     counter = "field" },
    { kind = "property_declaration",    handler = "field",  fn = property_text_free,  counter = "field" },
  }

  local INTERFACE_RULES = {
    { kind = "method_declaration",   handler = "method", fn = method_signature_free },
    { kind = "property_declaration", handler = "field",  fn = property_text_free,   counter = "field" },
  }

  local function extract_class_like(node, source, section, keyword)
    local mods = modifiers_text_free(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local bases = base_list_text(node, source)
    local text = compact_ws(prefixed(mods, keyword .. " " .. name .. " " .. bases))
    local entry = new_entry(section, node, text)
    local body = find_child(node, "declaration_list")
    if body then
      entry.children = extract_body_members(body, source, CLASS_RULES)
    end
    return entry
  end

  local function extract_interface(node, source)
    local mods = modifiers_text_free(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local bases = base_list_text(node, source)
    local text = compact_ws(prefixed(mods, "interface " .. name .. bases))
    local entry = new_entry(SECTION.Trait, node, text)
    local body = find_child(node, "declaration_list")
    if body then
      entry.children = extract_body_members(body, source, INTERFACE_RULES)
    end
    return entry
  end

  local function extract_enum(node, source)
    local mods = modifiers_text_free(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local text = compact_ws(prefixed(mods, "enum " .. name))
    local entry = new_entry(SECTION.Type, node, text)
    local body = find_child(node, "enum_member_declaration_list")
    if body then
      entry.children = extract_enum_variants(body, source, "enum_member_declaration")
      entry.child_kind = CHILD_BRIEF
    end
    return entry
  end

  local function extract_record(node, source)
    local mods = modifiers_text_free(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local params_node = find_child(node, "parameter_list")
    local params = params_node and get_text(params_node, source) or ""
    local bases = base_list_text(node, source)
    local text = compact_ws(prefixed(mods, "record " .. name .. params .. bases))
    return new_entry(SECTION.Type, node, text)
  end

  local function extract_namespace(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    return new_entry(SECTION.Module, node, get_text(name_node, source))
  end

  return {
    import_separator = ".",

    is_doc_comment = function(node, source)
      local kind = node:type()
      if kind == "single_line_doc_comment" then return true end
      if kind == "comment" then
        return get_text(node, source):sub(1, 3) == "///"
      end
      return false
    end,

    extract_nodes = function(node, source, _attrs)
      local kind = node:type()
      if kind == "using_directive" then
        return { simple_import(node, source, {"using%s+"}, ".") }
      elseif kind == "namespace_declaration" or kind == "file_scoped_namespace_declaration" then
        local e = extract_namespace(node, source)
        return e and { e } or {}
      elseif kind == "class_declaration" then
        local e = extract_class_like(node, source, SECTION.Class, "class")
        return e and { e } or {}
      elseif kind == "struct_declaration" then
        local e = extract_class_like(node, source, SECTION.Type, "struct")
        return e and { e } or {}
      elseif kind == "interface_declaration" then
        local e = extract_interface(node, source)
        return e and { e } or {}
      elseif kind == "enum_declaration" then
        local e = extract_enum(node, source)
        return e and { e } or {}
      elseif kind == "record_declaration" then
        local e = extract_record(node, source)
        return e and { e } or {}
      end
      return {}
    end,
  }
end
